src/linker/Linker.Steps/UnseenCallerAnnotateStep.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+using System.Runtime.CompilerServices;
+
+namespace Mono.Linker.Steps {
+	public class UnseenCallerAnnotateStep : BaseStep
+	{
+		AssemblyDefinition assembly;
+		MethodReference noOptAttr;
+
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			if (Annotations.GetAction (assembly) != AssemblyAction.Link)
+				return;
+
+			this.assembly = assembly;
+
+			foreach (var type in assembly.MainModule.Types)
+				ProcessType (type);
+		}
+
+		void ProcessType (TypeDefinition type)
+		{
+			foreach (var method in type.Methods) {
```

Comment:
If all of the private methods can have this attribute, I now hoist it and apply it to the type.

Reply:
The meaning of the attribute is to completely block reflection, not just the private methods on a type - there's nothing in the attribute's name implying it would be for privates only. .NET Native and CoreRT will discard metadata completely when this attribute is applied on a type (we get like [30%](https://github.com/dotnet/corert/pull/3685#issuecomment-303579581) size on disk savings from reflection blocking in general, so aggressively discarding metadata is pretty huge).

There is an attribute with the semantic we would want for this: [DisablePrivateReflectionAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.disableprivatereflectionattribute?view=netframework-4.7.2). I would suggest using that. I know the AttributeUsage doesn't allow using it on types, but tools typically don't care. We could file a CoreFX issue to allow extending the AttributeUsage.

Reply:
MichalStrehovsky Did that here:  https://github.com/mono/linker/pull/506/commits/7dfa1ecd4fe446792ad2e0e5779f169be02d02f3